driver/src/price_estimation/average_price_source.rs

```diff
@@ -2,89 +2,113 @@ use super::{PriceSource, Token};
 use crate::models::TokenId;
 use anyhow::{anyhow, Result};
 use futures::future::{self, BoxFuture, FutureExt as _};
-use std::collections::HashMap;
+use std::collections::{HashMap, HashSet};
+use std::hash::Hash;
 
-/// Combines two prices into one average.
-pub struct AveragePriceSource<T0, T1> {
-    source_0: T0,
-    source_1: T1,
+pub struct AveragePriceSource {
+    sources: Vec<Box<dyn PriceSource + Send + Sync>>,
 }
 
-impl<T0, T1> AveragePriceSource<T0, T1> {
-    pub fn new(source_0: T0, source_1: T1) -> Self {
-        Self { source_0, source_1 }
+impl AveragePriceSource {
+    pub fn new(sources: Vec<Box<dyn PriceSource + Send + Sync>>) -> Self {
+        Self { sources }
     }
 }
 
-impl<T0, T1> PriceSource for AveragePriceSource<T0, T1>
-where
-    T0: PriceSource + Send + Sync,
-    T1: PriceSource + Send + Sync,
-{
+impl PriceSource for AveragePriceSource {
     fn get_prices<'a>(
         &'a self,
         tokens: &'a [Token],
     ) -> BoxFuture<'a, Result<HashMap<TokenId, u128>>> {
         async move {
-            let prices = future::join(
-                self.source_0.get_prices(tokens),
-                self.source_1.get_prices(tokens),
-            )
-            .await;
-            match prices {
-                (Ok(p0), Ok(p1)) => Ok(average_prices(p0, p1)),
-                (Ok(p), Err(e)) | (Err(e), Ok(p)) => {
-                    log::warn!("one price source failed: {}", e);
-                    Ok(p)
-                }
-                (Err(e0), Err(e1)) => Err(anyhow!("both price sources failed: {}, {}", e0, e1)),
+            let price_futures =
+                future::join_all(self.sources.iter().map(|s| s.get_prices(tokens))).await;
+
+            let acquired_prices: Vec<HashMap<TokenId, u128>> = price_futures
+                .into_iter()
+                .filter_map(|f| match f {
+                    Ok(prices) => Some(prices),
+                    Err(err) => {
+                        log::warn!("Price source failed: {}", err);
+                        None
+                    }
+                })
+                .collect();
+
+            if !acquired_prices.is_empty() {
+                Ok(average_prices(acquired_prices))
+            } else {
+                Err(anyhow!("All price sources failed!"))
             }
         }
         .boxed()
     }
 }
 
-fn average_prices(
-    prices_0: HashMap<TokenId, u128>,
-    mut prices_1: HashMap<TokenId, u128>,
-) -> HashMap<TokenId, u128> {
-    for (token_id, price_1) in prices_0 {
-        prices_1
-            .entry(token_id)
-            .and_modify(|price| *price = (*price + price_1) / 2)
-            .or_insert(price_1);
+/// Lossless merger of a collection of maps puts all available values into a list for each available key
+pub fn lossless_merge<T: Copy + Eq + Hash + PartialEq, U: Clone + Copy>(
```

Comment:
ubernit: Because you are passing in the `map_collection` by value, you should be able to implement this without `U: Clone + Copy`. It would involve using `into_iter` instead of `iter` to get an interator on owned `U`s.

Since we are only using this for `u128`s here, this isn't an issue. But it would prevent this method from being used by types where `U: !Copy`.

Reply:
So I tried to remove the Clone and Copy from U, but then had to remove a copied inside the filter map and now somehow wind up with something like 

```
a value of type `std::vec::Vec<U>` cannot be built from an iterator over elements of type `&U`
```

Reply:
You might have to use `into_iter` instead of `iter` and `remove` instead of `get`, since you need to transfer ownership.